### ISSUE TITLE
bernoulli samples fail for small probabilities

### ISSUE THREAD

**Issue by jakevdp** [jax-ml/jax#28017]
Example:
```python
import jax

key = jax.random.key(0)
samples = jax.random.bernoulli(key, p=1E-10, shape=int(1E8))
print(samples.sum())  # should be zero with high probability
# 7
```
The issue is that at float32, uniform values have a resolution of $2^{-23}$ which is about $10^{-7}$, and so we don't have the resolution to properly evaluate samples with `p` smaller than about `1E-7`.

We could make this better by implementing bernoulli sampling directly on uint32 bits rather than converting to float, though that would only get us to `p` of about `2E-10`. By implementing an optional super-sampling (i.e. splitting the key) we could get to `4E-10`.

**Comment by Aniketsy** [comment]
@jakevdp 
https://github.com/jax-ml/jax/blob/main/jax/experimental/sparse/bcoo.py  

Sir in this I want to ask that. if we only need to check if indices are sorted and if not sorted we need to sort .
I’d really appreciate any feedback or suggestions to improve this approach

def _unique_indices_unbatched(indices, *, shape, return_inverse=False,
                              return_index=False, return_true_size=False):
  props = _validate_bcoo_indices(indices, shape)
  if props.n_sparse == 0:
    nse = 1
    indices_out = jnp.zeros_like(indices, shape=(nse, 0))
    out = (indices_out,)
    if return_index:
      out = (*out, jnp.zeros(nse, dtype='int32'))
    if return_inverse:
      out = (*out, jnp.zeros(nse, dtype='int32'))
    if return_true_size:
      out = (*out, nse)
    return out[0] if len(out) == 1 else out
  fill_value = jnp.expand_dims(jnp.array(shape[:props.n_sparse], dtype=indices.dtype), (0,))
  out_of_bounds = (indices >= fill_value).any(-1, keepdims=True)
  indices = jnp.where(out_of_bounds, fill_value, indices)
  # TODO: check if `indices_sorted` is True.
  out = _unique(indices, axis=0, return_inverse=return_inverse, return_index=return_index,
                return_true_size=return_true_size, size=props.nse, fill_value=fill_value)
  if return_inverse:
    idx = 2 if return_index else 1
    out = (*out[:idx], out[idx].ravel(), *out[idx + 1:])
  if return_true_size:
    nse = out[-1]
    nse = nse - (indices == fill_value).any().astype(nse.dtype)
    out = (*out[:-1], nse)
  return out


# This will work for 1D arrays
indices_sorted = jnp.all(jnp.diff(indices) >= 0)

if not indices_sorted:
      indices  = jnp.sort(indices, axis = 0)

**Comment by jakevdp** [comment]
Re-opening this, because #28022 had to be rolled back
